tests/core/arch-x86/stalker-x86.c

```diff
@@ -375,6 +401,239 @@ invoke_flat (TestStalkerFixture * fixture,
+  g_hash_table_destroy (compiled_run2);
+  g_hash_table_destroy (executed_run2);
+  g_hash_table_destroy (compiled_run1);
+  g_hash_table_destroy (executed_run1);
+
+  gum_stalker_unfollow_me (fixture->stalker);
+
+  g_print ("DONE\n");
+
+}
+
+static void
+prefetch_transformer (GumStalkerIterator * iterator,
+                   GumStalkerOutput * output,
+                   gpointer user_data)
+{
+  const cs_insn * instr;
+  gboolean begin = TRUE;
+  gpointer block_address;
+  gpointer * callout_ctx;
+
+  while (gum_stalker_iterator_next (iterator, &instr))
+  {
+    if (begin)
+    {
+      block_address = GSIZE_TO_POINTER (instr->address);
+      callout_ctx = gum_malloc (sizeof (gpointer));
+      *callout_ctx = block_address;
+
+      gum_stalker_iterator_put_callout (iterator, prefetch_callout,
+          callout_ctx, gum_free);
+
+      if (prefetch_compiled != NULL)
+      {
+        g_hash_table_insert (prefetch_compiled, block_address, block_address);
+      }
+      begin = FALSE;
+    }
+
+    gum_stalker_iterator_keep (iterator);
+  }
+}
+
+static void
+prefetch_callout (GumCpuContext * cpu_context,
+                  gpointer user_data)
+{
+  gpointer * callout_ctx = user_data;
+  gpointer block_address = *callout_ctx;
+
+  if (prefetch_executed != NULL)
+  {
+    g_hash_table_insert (prefetch_executed, block_address, block_address);
+  }
+}
+
+static void
+prefetch_run_child (GumStalker * stalker,
+                    GumMemoryRange * runner_range,
+                    int compile_fd,
+                    int execute_fd)
+{
+  pid_t pid;
+  int ret;
+  int status;
+
+  pid = fork ();
+  g_assert_cmpint (pid, >=, 0);
+  if (pid == 0)
+  {
+    /* child */
+    prefetch_compiled = g_hash_table_new (NULL, NULL);
+    prefetch_executed = g_hash_table_new (NULL, NULL);
+
+    gum_stalker_activate (stalker, prefetch_activation_target);
+    prefetch_activation_target ();
+    pretend_workload (runner_range);
+    gum_stalker_unfollow_me (stalker);
+
+    prefetch_write_blocks (compile_fd, prefetch_compiled);
+    prefetch_write_blocks (execute_fd, prefetch_executed);
+
+    g_hash_table_destroy (prefetch_compiled);
+    prefetch_compiled = NULL;
+    g_hash_table_destroy (prefetch_executed);
+    prefetch_executed = NULL;
+
+    exit (0);
+    return;
+  }
+
+  /* Wait for the child */
+  ret = waitpid (pid, &status, 0);
+  g_assert_cmpint (ret, ==, pid);
+  g_assert_cmpint (WIFEXITED (status), !=, 0);
+  g_assert_cmpint (WEXITSTATUS (status), ==, 0);
+}
+
+__attribute__((always_inline))
+static inline void
+prefetch_activation_target (void)
+{
+
+}
+
+static void
+prefetch_write_blocks (int fd,
+                       GHashTable * table)
+{
+  GHashTableIter iter;
+  gpointer iter_key, iter_value;
+  int ret;
+
+  g_hash_table_iter_init (&iter, table);
+  while (g_hash_table_iter_next (&iter, &iter_key, &iter_value))
+  {
+    ret = write (fd, &iter_key, sizeof (gpointer));
+    g_assert_cmpint (ret, ==, sizeof (gpointer));
```

Comment:
```suggestion
    g_assert_cmpint (ret, ==, sizeof (gpointer));
```

May fail because of partial writes (if the buffer is full), and theoretically also `EINTR`, so might be good to handle both in case users copy this example and use it in production code.

Reply:
Partial writes can be mitigated by specifying O_DIRECT in pipe2(), but I think it makes each message take up a whole page? man fcntl says:

"""

              When allocating the buffer for the pipe, the kernel may
              use a capacity larger than arg, if that is convenient for
              the implementation.  (In the current implementation, the
              allocation is the next higher power-of-two page-size
              multiple of the requested size.)  The actual capacity (in
              bytes) that is set is returned as the function result.
"""

So it seems unlikely we would hit a partial write from a full buffer? 

We should only get EINTR if the syscall is interrupted by signal delivery, we wouldn't expect this in the test harness. But it should be easy enough to handle, so let's just do that.

Reply:
Should I add the comment about partial writes to the code?

Reply:
Yeah actually let's just ignore it and not comment it – shouldn't worry about folks expecting too much from tests, as they're allowed to make a lot of assumptions.